geom/twkb_write.go

```diff
@@ -0,0 +1,535 @@
+	numPoints := coords.Length()
+	if !w.closeRings && numPoints >= 2 {
+		numPoints-- // Omit the final point in the ring.
+	}
+	w.writeUnsignedVarint(uint64(numPoints))
+	w.writePointArray(numPoints, coords.floats)
+	return nil
+}
+
+func (w *TWKBWriter) writePolygon(poly Polygon) error {
+	w.writeTypeAndPrecision(twkbTypePolygon)
+
+	if ctype := poly.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if poly.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	return w.writePolygonRings(poly)
+}
+
+func (w *TWKBWriter) writePolygonRings(poly Polygon) error {
+	w.writeUnsignedVarint(uint64(poly.NumRings()))
+
+	if poly.NumRings() == 0 {
+		return nil
+	}
+
+	ls := poly.ExteriorRing()
+	w.writeRing(ls)
+
+	numRings := poly.NumInteriorRings()
+	for i := 0; i < numRings; i++ {
+		ls = poly.InteriorRingN(i)
+		w.writeRing(ls)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiPoint(mp MultiPoint) error {
+	w.writeTypeAndPrecision(twkbTypeMultiPoint)
+
+	if ctype := mp.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if mp.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numPoints := mp.NumPoints()
+	w.writeUnsignedVarint(uint64(numPoints))
+
+	if err := w.writeIDList(numPoints); err != nil {
+		return err
+	}
+
+	for i := 0; i < numPoints; i++ {
+		pt := mp.PointN(i)
+		w.writePointCoords(pt)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiLineString(ml MultiLineString) error {
+	w.writeTypeAndPrecision(twkbTypeMultiLineString)
+
+	if ctype := ml.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if ml.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numLineStrings := ml.NumLineStrings()
+	w.writeUnsignedVarint(uint64(numLineStrings))
+
+	if err := w.writeIDList(numLineStrings); err != nil {
+		return err
+	}
+
+	for i := 0; i < numLineStrings; i++ {
+		ls := ml.LineStringN(i)
+		w.writeLineStringCoords(ls)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeMultiPolygon(mp MultiPolygon) error {
+	w.writeTypeAndPrecision(twkbTypeMultiPolygon)
+
+	if ctype := mp.CoordinatesType(); ctype != w.ctype {
+		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
+	}
+
+	if mp.IsEmpty() {
+		w.writeIsEmptyHeader()
+		return nil
+	}
+	w.writeInitialHeaders()
+
+	numPolygons := mp.NumPolygons()
+	w.writeUnsignedVarint(uint64(numPolygons))
+
+	if err := w.writeIDList(numPolygons); err != nil {
+		return err
+	}
+
+	for i := 0; i < numPolygons; i++ {
+		poly := mp.PolygonN(i)
+		w.writePolygonRings(poly)
+	}
+	return nil
+}
+
+func (w *TWKBWriter) writeGeometryCollection(gc GeometryCollection) error {
+	w.writeTypeAndPrecision(twkbTypeGeometryCollection)
```

Comment:
It seems (I could be wrong) that each geom's write method has a common block of code to start with:

```
	w.writeTypeAndPrecision(<the type enum>)

	if ctype := gc.CoordinatesType(); ctype != w.ctype {
		return fmt.Errorf("mismatched coordinate dimensions got %s expected %s", ctype, w.ctype)
	}

	if gc.IsEmpty() {
		w.writeIsEmptyHeader()
		return nil
	}
	w.writeInitialHeaders()
```

Perhaps this could be extracted out into a method?

Reply:
Although it's true there's common(ish) code at the top of each geometry type's write function, the behaviour of `IsEmpty()` for the various types may mean we want to treat them differently. I'll do some more testing before considering amalgamating any of this code.